### ISSUE TITLE
Ugly entries in command history

### ISSUE THREAD

**Issue by michael-heerdegen** [magit/transient#23]
Hello,

since some time, I've entries in my 'command-history' that look like

```
(magit-format-patch:--output-directory #s(#22=#s(eieio--class
transient-option #23="Class used for command-line argument that can
take a value." (#20=#s(eieio--class transient-argument #21="Abstract
superclass for infix arguments." (#17=#s(eieio--class transient-infix
#19="Transient infix command." (#14=#s(eieio--class transient-suffix
#16="Superclass for suffix command." (#2=#s(eieio--class
transient-child #3="Abstract superclass for group and and suffix
classes.

It is undefined what happens if more than one `if*' predicate slot is
non-nil." nil [#s(cl-slot-descriptor level 1 t ((:documentation
. #4="Enable if level of prefix is equal or greater.")))
#s(cl-slot-descriptor if nil t ((:documentation . #5="Enable if
predicate returns non-nil."))) #s(cl-slot-descriptor if-not nil t
((:documentation . #6="Enable if predicate returns nil.")))
#s(cl-slot-descriptor if-non-nil nil t ((:documentation . #7="Enable
if variable's value is non-nil."))) #s(cl-slot-descriptor if-nil nil t
((:documentation . #8="Enable if variable's value is nil.")))
#s(cl-slot-descriptor if-mode nil t ((:documentation . #9="Enable if
major-mode matches value."))) #s(cl-slot-descriptor if-not-mode nil t
((:documentation . #10="Enable if major-mode does not match value.")))
#s(cl-slot-descriptor if-derived nil t ((:documentation . #11="Enable
if major-mode derives from value."))) #s(cl-slot-descriptor
if-not-derived nil t ((:documentation . #12="Enable if major-mode does
not derive from value.")))] #s(hash-table size 65 test eq rehash-size
1.5 rehash-threshold 0.8125 data (level 0 if 1 if-not 2 if-non-nil 3
if-nil 4 if-mode 5 if-not-mode 6 if-derived 7 if-not-derived 8))
(transient-group transient-suffix) ((:level . level) (:if . if)
(:if-not . if-not) (:if-non-nil . if-non-nil) (:if-nil . if-nil)
(:if-mode . if-mode) (:if-not-mode . if-not-mode) (:if-derived
. if-derived) (:if-not-derived . if-not-derived)) #1=[] #1# #s(#2# 1
nil nil nil nil nil nil nil nil) (:custom-groups nil :documentation
#3# :abstract t))) [#s(cl-slot-descriptor level 1 t ((:documentation
. #4#))) #s(cl-slot-descriptor if nil t ((:documentation . #5#)))
#s(cl-slot-descriptor if-not nil t ((:documentation . #6#)))
#s(cl-slot-descriptor if-non-nil nil t ((:documentation . #7#)))
#s(cl-slot-descriptor if-nil nil t ((:documentation . #8#)))
#s(cl-slot-descriptor if-mode nil t ((:documentation . #9#)))
#s(cl-slot-descriptor if-not-mode nil t ((:documentation . #10#)))
#s(cl-slot-descriptor if-derived nil t ((:documentation . #11#)))
#s(cl-slot-descriptor if-not-derived nil t ((:documentation . #12#)))
#s(cl-slot-descriptor key #13=#:unbound t nil) #s(cl-slot-descriptor
command #13# t nil) #s(cl-slot-descriptor transient #13# t nil)
#s(cl-slot-descriptor format #15=" %k %d" t nil) #s(cl-slot-descriptor
description nil t nil)] #s(hash-table size 65 test eq rehash-size 1.5
rehash-threshold 0.8125 data (level 0 if 1 if-not 2 if-non-nil 3
if-nil 4 if-mode 5 if-not-mode 6 if-derived 7 if-not-derived 8 key 9
command 10 transient 11 format 12 description 13))
(magit--git-submodule-suffix transient-infix) ((:level . level) (:if
. if) (:if-not . if-not) (:if-non-nil . if-non-nil) (:if-nil . if-nil)
(:if-mode . if-mode) (:if-not-mode . if-not-mode) (:if-derived
. if-derived) (:if-not-derived . if-not-derived) (:key . key)
(:command . command) (:transient . transient) (:format . format)
(:description . description)) #1# #1# #s(#14# 1 nil nil nil nil nil
nil nil nil #13# #13# #13# #15# nil) (:custom-groups nil
:documentation #16#))) [#s(cl-slot-descriptor level 1 t
((:documentation . #4#))) #s(cl-slot-descriptor if nil t
((:documentation . #5#))) #s(cl-slot-descriptor if-not nil t
((:documentation . #6#))) #s(cl-slot-descriptor if-non-nil nil t
((:documentation . #7#))) #s(cl-slot-descriptor if-nil nil t
((:documentation . #8#))) #s(cl-slot-descriptor if-mode nil t
((:documentation . #9#))) #s(cl-slot-descriptor if-not-mode nil t
((:documentation . #10#))) #s(cl-slot-descriptor if-derived nil t
((:documentation . #11#))) #s(cl-slot-descriptor if-not-derived nil t
((:documentation . #12#))) #s(cl-slot-descriptor key #13# t nil)
#s(cl-slot-descriptor command #13# t nil) #s(cl-slot-descriptor
transient t t nil) #s(cl-slot-descriptor format #18=" %k %d (%v)" t
nil) #s(cl-slot-descriptor description nil t nil)
#s(cl-slot-descriptor argument #13# t nil) #s(cl-slot-descriptor
shortarg #13# t nil) #s(cl-slot-descriptor value nil t nil)
#s(cl-slot-descriptor multi-value nil t nil) #s(cl-slot-descriptor
allow-empty nil t nil) #s(cl-slot-descriptor history-key nil t nil)
#s(cl-slot-descriptor reader nil t nil) #s(cl-slot-descriptor prompt
nil t nil) #s(cl-slot-descriptor choices nil t nil)] #s(hash-table
size 65 test eq rehash-size 1.5 rehash-threshold 0.8125 data (level 0
if 1 if-not 2 if-non-nil 3 if-nil 4 if-mode 5 if-not-mode 6 if-derived
7 if-not-derived 8 key 9 command 10 transient 11 format 12 description
13 argument 14 shortarg 15 value 16 multi-value 17 allow-empty 18
history-key 19 reader 20 prompt 21 choices 22)) (transient-files
transient-variable transient-argument) ((:level . level) (:if . if)
(:if-not . if-not) (:if-non-nil . if-non-nil) (:if-nil . if-nil)
(:if-mode . if-mode) (:if-not-mode . if-not-mode) (:if-derived
. if-derived) (:if-not-derived . if-not-derived) (:key . key)
(:command . command) (:transient . transient) (:format . format)
(:description . description) (:argument . argument) (:shortarg
. shortarg) (:multi-value . multi-value) (:allow-empty . allow-empty)
(:history-key . history-key) (:reader . reader) (:prompt . prompt)
(:choices . choices)) #1# #1# #s(#17# 1 nil nil nil nil nil nil nil
nil #13# #13# t #18# nil #13# #13# nil nil nil nil nil nil nil)
(:custom-groups nil :documentation #19# :abstract t)))
[#s(cl-slot-descriptor level 1 t ((:documentation . #4#)))
#s(cl-slot-descriptor if nil t ((:documentation . #5#)))
#s(cl-slot-descriptor if-not nil t ((:documentation . #6#)))
#s(cl-slot-descriptor if-non-nil nil t ((:documentation . #7#)))
#s(cl-slot-descriptor if-nil nil t ((:documentation . #8#)))
#s(cl-slot-descriptor if-mode nil t ((:documentation . #9#)))
#s(cl-slot-descriptor if-not-mode nil t ((:documentation . #10#)))
#s(cl-slot-descriptor if-derived nil t ((:documentation . #11#)))
#s(cl-slot-descriptor if-not-derived nil t ((:documentation . #12#)))
#s(cl-slot-descriptor key #13# t nil) #s(cl-slot-descriptor command
#13# t nil) #s(cl-slot-descriptor transient t t nil)
#s(cl-slot-descriptor format #18# t nil) #s(cl-slot-descriptor
description nil t nil) #s(cl-slot-descriptor argument #13# t nil)
#s(cl-slot-descriptor shortarg #13# t nil) #s(cl-slot-descriptor value
nil t nil) #s(cl-slot-descriptor multi-value nil t nil)
#s(cl-slot-descriptor allow-empty nil t nil) #s(cl-slot-descriptor
history-key nil t nil) #s(cl-slot-descriptor reader nil t nil)
#s(cl-slot-descriptor prompt nil t nil) #s(cl-slot-descriptor choices
nil t nil)] #s(hash-table size 65 test eq rehash-size 1.5
rehash-threshold 0.8125 data (level 0 if 1 if-not 2 if-non-nil 3
if-nil 4 if-mode 5 if-not-mode 6 if-derived 7 if-not-derived 8 key 9
command 10 transient 11 format 12 description 13 argument 14 shortarg
15 value 16 multi-value 17 allow-empty 18 history-key 19 reader 20
prompt 21 choices 22)) (transient-switches transient-option
transient-switch) ((:level . level) (:if . if) (:if-not . if-not)
(:if-non-nil . if-non-nil) (:if-nil . if-nil) (:if-mode . if-mode)
(:if-not-mode . if-not-mode) (:if-derived . if-derived)
(:if-not-derived . if-not-derived) (:key . key) (:command . command)
(:transient . transient) (:format . format) (:description
. description) (:argument . argument) (:shortarg . shortarg)
(:multi-value . multi-value) (:allow-empty . allow-empty)
(:history-key . history-key) (:reader . reader) (:prompt . prompt)
(:choices . choices)) #1# #1# #s(#20# 1 nil nil nil nil nil nil nil
nil #13# #13# t #18# nil #13# #13# nil nil nil nil nil nil nil)
(:custom-groups nil :documentation #21# :abstract t)))
[#s(cl-slot-descriptor level 1 t ((:documentation . #4#)))
#s(cl-slot-descriptor if nil t ((:documentation . #5#)))
#s(cl-slot-descriptor if-not nil t ((:documentation . #6#)))
#s(cl-slot-descriptor if-non-nil nil t ((:documentation . #7#)))
#s(cl-slot-descriptor if-nil nil t ((:documentation . #8#)))
#s(cl-slot-descriptor if-mode nil t ((:documentation . #9#)))
#s(cl-slot-descriptor if-not-mode nil t ((:documentation . #10#)))
#s(cl-slot-descriptor if-derived nil t ((:documentation . #11#)))
#s(cl-slot-descriptor if-not-derived nil t ((:documentation . #12#)))
#s(cl-slot-descriptor key #13# t nil) #s(cl-slot-descriptor command
#13# t nil) #s(cl-slot-descriptor transient t t nil)
#s(cl-slot-descriptor format #18# t nil) #s(cl-slot-descriptor
description nil t nil) #s(cl-slot-descriptor argument #13# t nil)
#s(cl-slot-descriptor shortarg #13# t nil) #s(cl-slot-descriptor value
nil t nil) #s(cl-slot-descriptor multi-value nil t nil)
#s(cl-slot-descriptor allow-empty nil t nil) #s(cl-slot-descriptor
history-key nil t nil) #s(cl-slot-descriptor reader nil t nil)
#s(cl-slot-descriptor prompt nil t nil) #s(cl-slot-descriptor choices
nil t nil)] #s(hash-table size 65 test eq rehash-size 1.5
rehash-threshold 0.8125 data (level 0 if 1 if-not 2 if-non-nil 3
if-nil 4 if-mode 5 if-not-mode 6 if-derived 7 if-not-derived 8 key 9
command 10 transient 11 format 12 description 13 argument 14 shortarg
15 value 16 multi-value 17 allow-empty 18 history-key 19 reader 20
prompt 21 choices 22)) nil ((:level . level) (:if . if) (:if-not
. if-not) (:if-non-nil . if-non-nil) (:if-nil . if-nil) (:if-mode
. if-mode) (:if-not-mode . if-not-mode) (:if-derived . if-derived)
(:if-not-derived . if-not-derived) (:key . key) (:command . command)
(:transient . transient) (:format . format) (:description
. description) (:argument . argument) (:shortarg . shortarg)
(:multi-value . multi-value) (:allow-empty . allow-empty)
(:history-key . history-key) (:reader . reader) (:prompt . prompt)
(:choices . choices)) #1# #1# #s(#22# 1 nil nil nil nil nil nil nil
nil #13# #13# t #18# nil #13# #13# nil nil nil nil nil nil nil)
(:custom-groups nil :documentation #23#)) 1 nil nil nil nil nil nil
nil nil "C-m o " magit-format-patch:--output-directory t #18# "Output
directory" "--output-directory=" "-o" #24="/home/micha/today/" nil nil
nil transient-read-existing-directory nil nil) #24#)
```

Yes, that's one entry.  Such an entry is not useful to repeat a command because I can't say what is doing, it's just confusing.  Maybe you want to change things so that something more useful ends in the command history.

Thanks, Michael.

**Comment by tarsius** [comment]
It's unfortunate that this appears in `command-history` but I think this is the only downside.

Repeating an infix command is impossible while the transient is active (because transient doesn't allow you do run `repeat-complex-command`). Repeating the command again in most cases would not make any sense because the purpose of doing so is to set an argument; if there is nothing that consumes the argument then that won't do anything.

Before we continue lets look at what *every infix* command looks like:


```emacs-lisp
(lambda (obj value)
  (interactive
   (let ((obj (transient-suffix-object)))
     (list obj (transient-infix-read obj))))
  (transient-infix-set obj value)
  (transient--show))
```

For historic reasons some infix commands can be invoked directly. This only makes sense for infix commands that set a value by side-effect, e.g. `magit-remote.<remote>.url`. This only works if the command is explicitly taught how to do that (it needs to know the value for `<remote>`).

The term used for such a value is "scope", it can be thought of as some kind of "secondary value". The "primary value" or just "value" of a prefix command is the arguments that are being set by the infix arguments and passed to the suffix command. The secondary value is used for infix command that do not contribute to that value but instead set some external value by side-effect (in the above example the `remote.<remote>.url` Git variable.

When such an infix argument is invoked from a transient, then it gets the secondary value from the prefix. For example if you invoke the `magit-remote` transient, then the scope is set to the upstream of the current branch. If you invoke `magit-remote-configure` from `magit-remote`, then you obviously do so to configure *another* remote, so `magit-remote-configure` prompts for a remote.

Similarly if you invoke an argument-setting infix command directly it needs to ask you about the scope. This does have to be done before the value of the infix is read from the user (using `transient-infix-read` inside the commands `interactive` form).  We therefore do so inside `transient-suffix-object`.

The result is that the value (primary value) is passed to the function as an argument, but the scope (secondary value) is not.  That's a bit of a hack, but is acceptable when you consider that the infix commands are supposed to be invoked as, well, infix commands and that there use as stand-alone commands is only something that is provided for backward compatibility.

Now we could pass the scope as an additional argument, but then we would have to call `transient-suffix-object` inside the function body again.  That's a bit of a waste, but the bigger problem is that this function would now prompt for the scope again.  To avoid that we get the object once and pass it from `interactive` to the function body.

So in summary, I think this is not such a big problem because repeating an infix command isn't very useful and because the result is just some noise in `command-history`.

**Comment by michael-heerdegen** [comment]
Jonas Bernoulli <notifications@github.com> writes:

> Now we could pass the scope as an additional argument, but then we
> would have to call transient-suffix-object inside the function body
> again. That's a bit of a waste, but the bigger problem is that this
> function would now prompt for the scope again. To avoid that we get
> the object once and pass it from interactive to the function body.

There is another alternative: do nothing in the interactive form, move
all that to the body.  Would that be possible?  This is what I normally
try to do when I want to avoid clobbering command-history with nonsense.
Or the other way round: having some command adding junk to
command-history is often a sign that an interactive spec is returning
too much stuff that should go elsewhere.

> So in summary, I think this is not such a big problem because
> repeating an infix command isn't very useful and because the result is
> just some noise in command-history.

Yeah, it's just a detail.  But OTOH, if nobody would care,
command-history would become more and more unusable.  So it at least
makes some sense to try to fix it if it is possible.

Thanks, Michael.

**Comment by tarsius** [comment]
> There is another alternative: do nothing in the interactive form, move all that to the body.

Assuming that the primary reason why input should be read inside `interactive` whenever possible is to have it added to `command-history` and we want to *prevent* that, that seems like a reasonable thing to do.

**Comment by michael-heerdegen** [comment]
> Closed #23 via bbf8129.

Thanks!
